iceoryx_posh/test/moduletests/test_capro_service.cpp

```diff
@@ -28,166 +28,697 @@ using namespace iox::capro;
+    uint16_t testInstanceID2 = iox::capro::AnyInstance;
+    ServiceDescription serviceDescription1 = ServiceDescription(testServiceID1, testEventID1, testInstanceID1);
+    ServiceDescription serviceDescription2 = ServiceDescription(testServiceID2, testEventID2, testInstanceID2);
+
+    EXPECT_TRUE(serviceDescription1 == serviceDescription2);
+}
+
+TEST_F(ServiceDescription_test, TwoServiceDescriptionsWithDifferentButValidServiceIDsAreNotEqual)
+{
+    uint16_t testServiceID1 = 1U;
+    uint16_t testEventID1 = 2U;
+    uint16_t testInstanceID1 = 3U;
+    uint16_t testServiceID2 = 4U;
+    ServiceDescription serviceDescription1 = ServiceDescription(testServiceID1, testEventID1, testInstanceID1);
+    ServiceDescription serviceDescription2 = ServiceDescription(testServiceID2, testEventID1, testInstanceID1);
+
+    EXPECT_FALSE(serviceDescription1 == serviceDescription2);
+}
+
+TEST_F(ServiceDescription_test, TwoServiceDescriptionsWithDifferentButValidEventIDsAreNotEqual)
+{
+    uint16_t testServiceID1 = 1U;
+    uint16_t testEventID1 = 2U;
+    uint16_t testInstanceID1 = 3U;
+    uint16_t testEventID2 = 4U;
+    ServiceDescription serviceDescription1 = ServiceDescription(testServiceID1, testEventID1, testInstanceID1);
+    ServiceDescription serviceDescription2 = ServiceDescription(testServiceID1, testEventID2, testInstanceID1);
+
+    EXPECT_FALSE(serviceDescription1 == serviceDescription2);
+}
+
+TEST_F(ServiceDescription_test, TwoServiceDescriptionsWithDifferentButValidInstanceIDsAreNotEqual)
+{
+    uint16_t testServiceID1 = 1U;
+    uint16_t testEventID1 = 2U;
+    uint16_t testInstanceID1 = 3U;
+    uint16_t testInstanceID2 = 4U;
+    ServiceDescription serviceDescription1 = ServiceDescription(testServiceID1, testEventID1, testInstanceID1);
+    ServiceDescription serviceDescription2 = ServiceDescription(testServiceID1, testEventID1, testInstanceID2);
+
+    EXPECT_FALSE(serviceDescription1 == serviceDescription2);
+}
+
+TEST_F(ServiceDescription_test, TwoServiceDescriptionsWithDifferentAndValidServiceInstanceEventIDsAreNotEqual)
+{
+    uint16_t testServiceID1 = 1U;
+    uint16_t testEventID1 = 2U;
+    uint16_t testInstanceID1 = 3U;
+    uint16_t testServiceID2 = 4U;
+    uint16_t testEventID2 = 5U;
+    uint16_t testInstanceID2 = 6U;
+    ServiceDescription serviceDescription1 = ServiceDescription(testServiceID1, testEventID1, testInstanceID1);
+    ServiceDescription serviceDescription2 = ServiceDescription(testServiceID2, testEventID2, testInstanceID2);
+
+    EXPECT_FALSE(serviceDescription1 == serviceDescription2);
+}
+
+TEST_F(ServiceDescription_test, TwoServiceDescriptionsWithDifferentInstanceStringsAreNotEqual)
+{
+    uint16_t testServiceID = 1U;
+    uint16_t testEventID = 2U;
+    uint16_t testInstanceID = 3U;
+    testService = "1";
+    testEvent = "2";
+    testInstance = "instance";
+    ServiceDescription::ClassHash testHash = {1U, 2U, 3U, 4U};
+    auto serialObj = iox::cxx::Serialization::create(testService,
```

Comment:
Okay, I understand. But then, there shouldn't be a provision for the user to set it right ?

Reply:
Also to cover every branch in the == operator, serialization is needed.

Reply:
> Okay, I understand. But then, there shouldn't be a provision for the user to set it right ?
> Also to cover every branch in the == operator, serialization is needed.

The thing is that for a correct transfer of the ServiceDescription, all member variables need to be set by the Serialization. Currently i tend more into the direction that the Serialization expects a correct ServiceDescription as a contract. A clean solution would it be to prevent potential misuse.

In the comparison operator of the ServiceDescription are only members compared which can be set by the normal Ctors. Why is here the serialization needed?

The problem i see here is that you construct a ServiceDescription which has set internally `m_instanceString = "instance"` and `m_instanceID = 3` which does happen when you go the normal way with the constructor.
The serialization is here used as "backdoor" to inject the internal variables directly. This is not forbidden because it's public API but the serialization should normally not be used for that.
Keep in mind that also developers which are new to iceoryx look at tests to get an idea how the classes are used. They could get the wrong impression that the serialization is an alternative constructor, which we not want.

Maybe it's a good idea to set a doxygen tag `/// @attention` to make that visible.

Reply:
From my understanding,  its not possible to specifically set separate values for ID and the corresponding string because the constructor converts the string to integer to store it in the ID and vice versa to store it in the corresponding string. Nevertheless, for the branch coverage that I mentioned, Invalid ID case can be used to cover the branch and as you said, serialization could be removed. I'm new to this project and don't know much about it so please correct me if I'm wrong . :)

Reply:
> From my understanding, its not possible to specifically set separate values for ID and the corresponding string because the constructor converts the string to integer to store it in the ID and vice versa to store it in the corresponding string.

Yes correct, the "three friends" instance, service and event shall either be set as string or as unsiged integers. Mixed initialization is normally not allowed for the end-user. This is something which is not well documented and we need to improve on that.